modules/contact/src/constraints/ComputeFrictionalForceLMMechanicalContact.C

```diff
@@ -107,20 +113,59 @@ ComputeFrictionalForceLMMechanicalContact::residualSetup()
 }
 
 void
-ComputeFrictionalForceLMMechanicalContact::jacobianSetup()
+ComputeFrictionalForceLMMechanicalContact::communicateVelocities()
 {
-  residualSetup();
+#ifdef MOOSE_SPARSE_AD
+  // We may have weighted gap information that should go to other processes that own the dofs
+  using Datum = std::pair<dof_id_type, ADReal>;
+  std::unordered_map<processor_id_type, std::vector<Datum>> push_data;
+
+  for (auto & pr : _dof_to_weighted_tangential_velocity)
+  {
+    const auto * const dof_object = pr.first;
+    const auto proc_id = dof_object->processor_id();
+    if (proc_id == this->processor_id())
+      continue;
+
+    push_data[proc_id].push_back(std::make_pair(dof_object->id(), std::move(pr.second)));
+  }
+
+  const auto & lm_mesh = _mesh.getMesh();
+
+  auto action_functor = [this, &lm_mesh](const processor_id_type libmesh_dbg_var(pid),
+                                         const std::vector<Datum> & sent_data) {
+    mooseAssert(pid != this->processor_id(), "We do not send messages to ourself here");
```

Comment:
Send message to yourself is cheap, actually

Reply:
It may be cheap but the way I've setup the logic is that on the current process I don't setup pushes to myself so I want to make sure that I'm actually doing what I think I'm doing